[FEATURE]: Generics, lifetimes, where clauses and internal field variant formatting

jngls opened the pull request:
This is a considerable update of the proc macro which supports optional custom format strings per variant with full access to variant fields via the `#[variant_display(format = "xyz")]` attribute.

I recommend reading the code in isolation rather than as a diff.

# Summary
* Added `EnumAttrs` which handles parsing of the main `enum_display` attribute
* Added `VariantAttrs` which handles parsing of the new `variant_display` field attribute
* `variant_display` field attribute allows the user to customise each variant
  - The `format = "xyz"` argument defines a custom display format (see below for **Format Behaviour**) - if omitted, the original crate behaviour is used
* Introduced a new Intermediate Representation (IR) to hold variant info and allow for more complex processing:
  - `VariantInfo` holds properties common to all variant types
  - `NamedVariantIR`, `UnnamedVariantIR`, and `UnitVariantIR` are the intermediate representations of variant types
  - `VariantIR` wraps all of those
* Pipeline updated to account for new Intermediate Representation
  - Parsing of `syn` tokens is moved to the `from_*()` methods of `EnumAttrs`, `VariantAttrs`, and `VariantIR`
  - We detect if any variant specifies a "format" string - this is to ensure all match arms return the same type
  - Finally we generate match arm tokens with the `VariantIR::gen()` method
* Tests have been updated

# Format Behaviour
Format strings behave almost identically to using `format!()` in normal rust code so it should be very familiar to users. Besides a bit of translation of unnamed fields, the string specified by the user is the string that ends up as the parameter to `format!()` in each match arm.

The following considerations apply:
* To print the case adjusted variant name, you use the `{variant}` placeholder
* To print named variant fields, you use the field name placeholder, eg. if we have `Variant { my_field: u32 }`, you use the `{my_field}` placeholder
* To print unnamed variant fields, you use a numbered placeholder, eg. if we have `Variant(u32)`, you use the `{0}` placeholder (to access the first tuple element)
* It is not currently possible to access fields within fields - each field must implment its own Display or Debug
* Most of the common inline format specs are supported, eg. `{my_field:?}` displays `my_field` using its `Debug` implementation, `{0:5}` displays the first unnamed field with a width of 5 - however - format spec options which depend on extra arguments passed into `format!()` are not supported

# Additional Notes
* I haven't included any doc changes in this PR - if you like the PR, then I'll happily update the docs
* I've left it to the user to decide if their target platform supports `format!()` - perhaps instead, we should [extern alloc](https://stackoverflow.com/questions/71675411/refer-to-an-extern-crate-in-macro-expansion) and hide it behind a feature flag?
* This branch accidentally includes code from #4 as that was small and done in `main` - let me know if you want them separated. Otherwise merging #4 first or discarding and just merging this would also work.

# Full examples
```rust
#[derive(EnumDisplay)]
enum TestEnum {
    // Displayed as "Name"
    Name,

    // Displayed as "Unit: NameFullFormat"
    #[variant_display(format = "Unit: {variant}")]
    NameFullFormat,

    // Displayed as "Address"
    Address {
        street: String,
        city: String,
        state: String,
        zip: String,
    },

    // Displayed as "Named: AddressPartialFormat {123 Main St, 12345}"
    #[variant_display(format = "Named: {variant} {{{street}, {zip}}}")]
    AddressPartialFormat {
        street: String,
        city: String,
        state: String,
        zip: String,
    },

    // Displayed as "Named: AddressFullFormat {123 Main St, Any Town, CA, 12345}"
    #[variant_display(format = "Named: {variant} {{{street}, {city}, {state}, {zip}}}")]
    AddressFullFormat {
        street: String,
        city: String,
        state: String,
        zip: String,
    },

    // Displayed as "DateOfBirth"
    DateOfBirth(u32, u32, u32),

    // Displayed as "Unnamed: DateOfBirthPartialFormat(1999)"
    #[variant_display(format = "Unnamed: {variant}({2})")]
    DateOfBirthPartialFormat(u32, u32, u32),

    // Displayed as "Unnamed: DateOfBirthFullFormat(1, 2, 1999)"
    #[variant_display(format = "Unnamed: {variant}({0}, {1}, {2})")]
    DateOfBirthFullFormat(u32, u32, u32),
}

#[test]
fn test_unit_field_variant() {
    assert_eq!(TestEnum::Name.to_string(), "Name");
    assert_eq!(TestEnum::NameFullFormat.to_string(), "Unit: NameFullFormat");
}

#[test]
fn test_named_fields_variant() {
    assert_eq!(
        TestEnum::Address {
            street: "123 Main St".to_string(),
            city: "Any Town".to_string(),
            state: "CA".to_string(),
            zip: "12345".to_string()
        }
        .to_string(),
        "Address"
    );
    assert_eq!(
        TestEnum::AddressPartialFormat {
            street: "123 Main St".to_string(),
            city: "Any Town".to_string(),
            state: "CA".to_string(),
            zip: "12345".to_string()
        }
        .to_string(),
        "Named: AddressPartialFormat {123 Main St, 12345}"
    );
    assert_eq!(
        TestEnum::AddressFullFormat {
            street: "123 Main St".to_string(),
            city: "Any Town".to_string(),
            state: "CA".to_string(),
            zip: "12345".to_string()
        }
        .to_string(),
        "Named: AddressFullFormat {123 Main St, Any Town, CA, 12345}"
    );
}

#[test]
fn test_unnamed_fields_variant() {
    assert_eq!(TestEnum::DateOfBirth(1, 2, 1999).to_string(), "DateOfBirth");
    assert_eq!(TestEnum::DateOfBirthPartialFormat(1, 2, 1999).to_string(), "Unnamed: DateOfBirthPartialFormat(1999)");
    assert_eq!(TestEnum::DateOfBirthFullFormat(1, 2, 1999).to_string(), "Unnamed: DateOfBirthFullFormat(1, 2, 1999)");
}
```

Commit:
Implement generics, lifetimes, bounds and where clause support

Commit:
First pass formating

Commit:
Simplify

jngls commented:
Btw, my specific use case was this:
```rust
#[derive(Debug, Clone, PartialEq, Eq, EnumDisplay)]
#[enum_display(case = "Lower")]
pub enum ParserErrorKind<I> {
    NomError(ErrorKind),
    #[variant_display(format = "{variant}: {0}")]
    Keyword(Keyword),
    Ident(Ident<I>),
    ...
}
```

Where I wanted to provide a general impl of Display for most variants, but be specific for Keyword.

SeedyROM commented:
Thanks for submitting these PRs! Both the generics impl and this. I'll have time next weekend to actually look at this and critique. From the 1000 mile overview it looks like a solid addition to this library.

It seems like the build pipeline is currently failing and I'll investigate that more as well.

Thanks again!

jngls commented:
Cheers! Glad you like it on the surface.

The generics side of things was actually really simple (almost like the code equivalent of copy paste - you take the generics from `syn` and forward them on), so I'd be surprised if that was at fault.

I've had a quick look at the build issue and it's very weird. Does indeed look like maybe a version mismatch.

Anyway, yeah, like you, I'll be able to spend more time on this next weekend.

SeedyROM committed:
:hammer: attempt to fix issues

SeedyROM committed:
:rotating_light: Fix issues with current PR

- Bump version number before publish
- `gen` is a reserved keyword: https://doc.rust-lang.org/edition-guide/rust-2024/gen-keyword.html
- Rename all functions named `gen` to `generate`
- Allow for local testing via workspaces, however when publishing the local dep needs to be replaced with the published version
   - This allowed for the actual proc_macro code changes to propegate into the local build.

SeedyROM committed:
:hammer: Use specific version locally, in production/package management use the published version

SeedyROM commented:
@jngls I created a branch on top of your changes you pushed here and have got the CI/CD pipeline building and fixed a few other things here: https://github.com/SeedyROM/enum-display/pull/6. You can see the fixes that were needed.

I'm unfortunately not a git master, so I think some of the history got borked with my changes being pushed upstream to my repo from your fork/branch. Feel free to add my commits to your PR and it works great if you want to properly preserve your impact on the codebase.

Thanks again for the great code and suggestions here. I hope my fixes are good enough for you!

SeedyROM committed:
:hammer: API changes for PR

- Rename `variant_display` to `display`
- Support named argument `format=` and first argument for format.

SeedyROM committed:
:hammer: Update test cases to avoid using the explicit macro argument by default

SeedyROM committed:
:hammer: Update test name for case with lifetime

jngls commented:
Yo sorry, work was crazy lately. I finally have a day to myself, so I'll look at this now. Need to catch up on what you've done.

jngls commented:
Yeah, your changes look great to me. I was never happy with the long `variant_display` attribute name and only named it that way to be similar in format to `enum_display`. `display` works for me.

With the build issue, good fix. I remember now I had to locally change the dependency to `path` to get the tests to build.

I don't see anything borked in the git history. Just a bunch of autoformat, which is fine. Looks good to me. I'll merge your changes into my branch, test locally, and then start updating docs.

Commit:
:hammer: Update docs to round off formatting feature.

Commit:
:hammer: Clarity improvements in docs

jngls commented:
I've been unable to use:
```
quote! {
    #ident(#(#fields),*) => {
        #[cfg(feature = "std")]
        {
            let variant = #ident_transformed;
            format!(#fmt)
        }
        #[cfg(not(feature = "std"))]
        {
            #ident_transformed
        }
    }
}
```

This is because it mandates that the `std` feature is available in the user's crate, which we don't want.

However, I have a plan. The `std` feature will be available in `enum-display`, so we export a macro to wrap `format!()` and `String::from()` and that can just pass the raw variant string out for `no_std`.

It feels slightly hacky and I don't like the idea of exporting something the user isn't meant to see. But I'm not sure what else to do right now. Lemme know if you have any ideas. I'll push the code shortly.

Maybe this can serve as a first iteration and if we find a better method we can switch to that.

SeedyROM commented:
I never had any intention of supporting `no-std` actually, so if you can find a nice solution to add this feature that would be awesome! I support any ideas you have as you probably have more experience with it than me.

SeedyROM commented:
I was able to "vibe-code" (lol) a simple `no_std` solution for the proc macro to use `::core::write!()` instead of strings when possible.

```diff
diff --git a/enum-display-macro/src/lib.rs b/enum-display-macro/src/lib.rs
index 6483879..54628cb 100644
--- a/enum-display-macro/src/lib.rs
+++ b/enum-display-macro/src/lib.rs
@@ -150,9 +150,14 @@ impl NamedVariantIR {
         let fields = self.fields;
         match (any_has_format, attrs.format) {
             (true, Some(fmt)) => {
-                quote! { #ident { #(#fields),* } => { let variant = #ident_transformed; format!(#fmt) } }
+                quote! { #ident { #(#fields),* } => {
+                    let variant = #ident_transformed;
+                    ::core::write!(f, #fmt)
+                } }
+            }
+            (true, None) => {
+                quote! { #ident { .. } => ::core::fmt::Formatter::write_str(f, #ident_transformed), }
             }
-            (true, None) => quote! { #ident { .. } => String::from(#ident_transformed), },
             (false, None) => quote! { #ident { .. } => #ident_transformed, },
             _ => unreachable!(
                 "`any_has_format` should never be false when a variant has format string"
@@ -187,9 +192,14 @@ impl UnnamedVariantIR {
         let fields = self.fields;
         match (any_has_format, attrs.format) {
             (true, Some(fmt)) => {
-                quote! { #ident(#(#fields),*) => { let variant = #ident_transformed; format!(#fmt) } }
+                quote! { #ident(#(#fields),*) => {
+                    let variant = #ident_transformed;
+                    ::core::write!(f, #fmt)
+                } }
+            }
+            (true, None) => {
+                quote! { #ident(..) => ::core::fmt::Formatter::write_str(f, #ident_transformed), }
             }
-            (true, None) => quote! { #ident(..) => String::from(#ident_transformed), },
             (false, None) => quote! { #ident(..) => #ident_transformed, },
             _ => unreachable!(
                 "`any_has_format` should never be false when a variant has format string"
@@ -216,9 +226,14 @@ impl UnitVariantIR {
         } = self.info;
         match (any_has_format, attrs.format) {
             (true, Some(fmt)) => {
-                quote! { #ident => { let variant = #ident_transformed; format!(#fmt) } }
+                quote! { #ident => {
+                    let variant = #ident_transformed;
+                    ::core::write!(f, #fmt)
+                } }
+            }
+            (true, None) => {
+                quote! { #ident => ::core::fmt::Formatter::write_str(f, #ident_transformed), }
             }
-            (true, None) => quote! { #ident => String::from(#ident_transformed), },
             (false, None) => quote! { #ident => #ident_transformed, },
             _ => unreachable!(
                 "`any_has_format` should never be false when a variant has format string"
@@ -299,36 +314,44 @@ pub fn derive(input: TokenStream) -> TokenStream {
     .map(|variant| VariantIR::from_variant(variant, &enum_attrs))
     .collect();
 
-    // If any variants have a format string, the output of all match arms must be String instead of &str
-    // This is because we can't return a reference to the temporary output of format!()
+    // If any variants have a format string, we need to handle formatting differently
     let any_has_format = intermediate_variants.iter().any(|v| v.has_format());
-    let post_fix = if any_has_format {
-        quote! { .as_str() }
-    } else {
-        quote! {}
-    };
 
     // Build the match arms
     let variants = intermediate_variants
         .into_iter()
         .map(|v| v.generate(any_has_format));
 
-    // #[allow(unused_qualifications)] is needed
-    // due to https://github.com/SeedyROM/enum-display/issues/1
-    // Possibly related to https://github.com/rust-lang/rust/issues/96698
-    let output = quote! {
-        #[automatically_derived]
-        #[allow(unused_qualifications)]
-        impl #impl_generics ::core::fmt::Display for #ident #ty_generics #where_clause {
-            fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
-                ::core::fmt::Formatter::write_str(
-                    f,
+    let output = if any_has_format {
+        // When format strings are present, we write directly to the formatter
+        quote! {
+            #[automatically_derived]
+            #[allow(unused_qualifications)]
+            impl #impl_generics ::core::fmt::Display for #ident #ty_generics #where_clause {
+                fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
                     match self {
                         #(Self::#variants)*
-                    }#post_fix
-                )
+                    }
+                }
+            }
+        }
+    } else {
+        // When no format strings, we can return &str directly
+        quote! {
+            #[automatically_derived]
+            #[allow(unused_qualifications)]
+            impl #impl_generics ::core::fmt::Display for #ident #ty_generics #where_clause {
+                fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+                    ::core::fmt::Formatter::write_str(
+                        f,
+                        match self {
+                            #(Self::#variants)*
+                        }
+                    )
+                }
             }
         }
     };
+
     output.into()
 }
```

There's a few other configuration changes to the main lib tests:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 62ea829..11379d3 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,5 +1,8 @@
 //!
-//! enum-display is a crate for implementing [`std::fmt::Display`] on enum variants with macros.
+//! enum-display is a crate for implementing [`core::fmt::Display`] on enum variants with macros.
+//!
+//! This crate supports both `std` and `no_std` environments. In `no_std` mode, it works
+//! without allocation by writing directly to the formatter.
 //!
 //! # Simple Example
 //!
@@ -31,8 +34,28 @@
 //!     HelloGreeting { name: String },
 //! }
 //!
+//! # #[cfg(feature = "std")]
 //! assert_eq!(Message::HelloGreeting { name: "Alice".to_string() }.to_string(), "hello-greeting");
+//! ```
+//!
+//! # No-std Usage
+//!
+//! This crate works in `no_std` environments:
+//!
+//! ```rust
+//! # #![cfg_attr(not(feature = "std"), no_std)]
+//! use enum_display::EnumDisplay;
 //!
+//! #[derive(EnumDisplay)]
+//! enum Status {
+//!     Ready,
+//!     
+//!     #[display("Error: {code}")]
+//!     Error { code: u32 },
+//! }
+//! ```
+
+#![cfg_attr(not(feature = "std"), no_std)]
 
 pub use enum_display_macro::*;
 
@@ -40,6 +63,15 @@ pub use enum_display_macro::*;
 mod tests {
     use super::*;
 
+    #[cfg(feature = "std")]
+    use std::string::{String, ToString};
+
+    #[cfg(not(feature = "std"))]
+    extern crate alloc;
+
+    #[cfg(not(feature = "std"))]
+    use alloc::string::{String, ToString};
+
     #[allow(dead_code)]
     #[derive(EnumDisplay)]
     enum TestEnum {
@@ -101,7 +133,7 @@ mod tests {
     #[derive(EnumDisplay)]
     enum TestEnumWithLifetimeAndGenerics<'a, T: Clone>
     where
-        T: std::fmt::Display,
+        T: core::fmt::Display,
     {
         Name,
         Address {
```

Same with added features in `Cargo.toml`:

```
diff --git a/Cargo.toml b/Cargo.toml
index 1bf8a55..34c8250 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -14,3 +14,18 @@ repository = "https://github.com/SeedyROM/enum-display"
 
 [dependencies]
 enum-display-macro = { version = "0.1.5", path = "./enum-display-macro" }
+
+[features]
+default = ["std"]
+std = []
+
+[dev-dependencies]
+static_assertions = "1.1"
+
+[[test]]
+name = "no_std_integration"
+required-features = []
+
+[package.metadata.docs.rs]
+all-features = true
+rustdoc-args = ["--cfg", "docsrs"]
```

I think the most important changes are in the initial change to the proc macro. It also solves alot of the changes you made in your experimental branch I think??? I am dumb on this stuff but the tests are passing locally for me. We'll need more tests for embedded targets etc if we really want to verify.

jngls commented:
Eh. Whatever works. I think this is a good approach, at least from a surface skim read. But I'll have a deeper look when I can find the brain power.

I've got various embedded projects I can yank the setup from to test so I'll happily try it out. Main one to validate will be ARM family (eg. STM) since that's the most popular.

jngls committed:
Add diffs from SeedyROM as per 'https://github.com/SeedyROM/enum-display/pull/5#issuecomment-3146215371'

jngls commented:
Last thing that might be worth doing is updating the readme with the two new examples, but I'm inclined to leave that to you once it's submitted so you can format it the way you like.

SeedyROM commented:
@jngls  Shit I forgot to update the README.md and docs, but for now I actually need this feature in a product of mine. So we're gonna go with `0.2.0` for now and I'll update the docs in `0.2.1` if needed! I'm not a great open source maintainer so I'll take the hit on that. I jumped the gun lolololol.

jngls commented:
Hehe, all good. You're doing great! Will gladly work with you again.

Minor version bump makes sense for this.

SeedyROM commented:
Likewise! I owe you a virtual beer!
